q2_demux/_demux.py

```diff
@@ -111,17 +138,74 @@ def __iter__(self):
                 raise ValueError(
                     'Sequence header lines do not contain description fields '
                     'but barcode header lines do.')
-            elif _trim_trailing_slash(barcode_header.description) != \
-                    _trim_trailing_slash(sequence_header.description):
+            elif _trim_description(barcode_header.description) != \
+                    _trim_description(sequence_header.description):
                 raise ValueError(
                     'Mismatched sequence descriptions: %s and %s' %
-                    (_trim_trailing_slash(barcode_header.description),
-                     _trim_trailing_slash(sequence_header.description)))
+                    (_trim_description(barcode_header.description),
+                     _trim_description(sequence_header.description)))
 
             yield barcode_record, sequence_record
 
-    def __next__(self):
-        return next(self.barcode_generator), next(self.sequence_generator)
+
+class BarcodePairedSequenceFastqIterator(collections.abc.Iterable):
+    def __init__(self, barcode_generator, forward_generator,
+                 reverse_generator):
+        self.barcode_generator = barcode_generator
+        self.forward_generator = forward_generator
+        self.reverse_generator = reverse_generator
+
+    def __iter__(self):
+        # Adapted from q2-types
+        for barcode_record, forward_record, reverse_record \
+                in itertools.zip_longest(self.barcode_generator,
+                                         self.forward_generator,
+                                         self.reverse_generator):
+            if barcode_record is None:
+                raise ValueError('More sequences were provided than barcodes.')
+            if forward_record is None or reverse_record is None:
```

Comment:
Can you split this into two if statements so that the error message can indicate which file (forward or reverse) has too few reads? That will be really helpful for debugging when this comes up.